*The One* - nRF52 / embedded runner

daringer opened the pull request:
This is the long awaited multi MCU embedded runner featuring:

* `runners/embedded` to build for `nrf52` **and** for `lpc55` from based on a single code base
* `components/littlefs2` and `lpc55_hal` to implement some necessary compatibility with `nrf52_hal`   
* `lpc55` is still missing some parts to be built out of `embedded`
* and much, much more...

many thanks to @jnordholz who did most of the heavy lifting!

robin-nitrokey commented on .gitlab-ci.yml:
```diff
@@ -61,3 +61,28 @@ build-nk3am:
+    - rustup target add thumbv7em-none-eabihf
+    - mkdir -p artifacts
+    - make -C runners/embedded build-nk3am.bl FEATURES=provisioner
+    - cp runners/embedded/artifacts/*.bin artifacts/provisioner-firmware.bin
+    - cp runners/embedded/artifacts/*.ihex artifacts/provisioner-firmware.ihex
+    - make -C runners/embedded build-nk3am.bl FEATURES=develop
+    - cp runners/embedded/artifacts/*.bin artifacts/develop-firmware.bin
+    - cp runners/embedded/artifacts/*.ihex artifacts/develop-firmware.ihex
```
Can you please change the artifact filenames to match the lpc55 naming pattern?  Would be something like: `{firmware|provisioner}-nk3am-nrf52-{version}.zip`.

robin-nitrokey replied:
I forgot that these are the unsigned binaries so the exact name is not as important.  Still it would be good to add the soc and hardware version.

daringer replied:
done, taking version from Cargo.toml directly for now

robin-nitrokey commented on components/littlefs2/Cargo.toml:
```diff
@@ -0,0 +1,59 @@
+[package]
+name = "littlefs2"
```
Instead of copying the crate into this repository and modifying it, can we fork it and use it as a Git dependency?  Would be much easier to see and track the differences.

robin-nitrokey replied:
Same for `lpc55-hal`.

daringer replied:
moved `lpc55-hal` & `littlefs2` as forks into the Nitrokey org.

robin-nitrokey commented on components/provisioner-app/src/lib.rs:
```diff
@@ -31,7 +31,9 @@ use apdu_dispatch::iso7816::{Status, Instruction};
-use lpc55_hal as hal;
+#[cfg_attr(feature = "lpc55", path = "board_lpc55.rs")]
+#[cfg_attr(feature = "nrf52", path = "board_nrf52.rs")]
+mod board;
```
Can we move the UUID into the app struct and set it when constructing the app in the runner (similar to the admin app)?  That would reduce code duplication.  Similarly, we can provide a `Reboot` implementation for the reboot.

robin-nitrokey commented on runners/embedded/Cargo.toml:
```diff
@@ -0,0 +1,148 @@
+version = "0.0.1"
+authors = ["Jan Nordholz <j.nordholz@tu-berlin.de", "Markus Meissner <meissner@nitrokey.com>"]
+edition = "2018"
+license = "ISC"
```
Is the license choice intentional?  I think or consensus was to use GPL here, right?

daringer replied:
sticking to the repo-wide MIT for now

robin-nitrokey commented on runners/embedded/Makefile:
```diff
@@ -0,0 +1,277 @@
+
```
It would be good if we could simplify this Makefile a bit and move some of the nrf52-specific stuff somewhere else, but we don’t have to do this as part of this PR.

daringer replied:
for now moved out the bootloader stuff, needs more cleanups later
especially started to move dev & nrf specific things into `utils/`

robin-nitrokey commented on runners/embedded/Makefile.old:
```diff
@@ -0,0 +1,158 @@
+### NRF52 device serial ID <-> device type mapping.
```
Do we still need this old file?

daringer replied:
nope, trashed

robin-nitrokey reviewed (commented):
Great work!  I’ve added some comments based on my first impressions and will look at it more thoroughly later.

robin-nitrokey commented on runners/embedded/src/soc_lpc55/trussed.rs:
```diff
@@ -0,0 +1,209 @@
+}
+
+
+pub struct UserInterface<BUTTONS, RGB>
+where
+BUTTONS: Press + Edge,
+RGB: RgbLed,
+{
+    rtc: Rtc<init_state::Enabled>,
+    buttons: Option<BUTTONS>,
+    rgb: Option<RGB>,
+    wink: Option<core::ops::Range<Duration>>,
+    provisioner: bool,
+}
```
Can you pull in the changes from #51 (and eventually #53) before the tagging the release candidate so that we have consistent behavior on all devices?

daringer replied:
pulled them in, adapted the lpc55 and nrf52 specific parts respectively

robin-nitrokey commented on utils/nrf-bootloader/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+	build-essential python3 wget tar perl patch curl gdebi \
+	binutils-arm-none-eabi gcc-arm-none-eabi libnewlib-arm-none-eabi
+
+RUN wget https://github.com/nobodywasishere/nrf5-sdk/archive/v17.1.0-ddde560.tar.gz && \
```
https://github.com/Nitrokey/nitrokey-3-firmware/issues/41 asked to use an official release here.

daringer replied:
The issue is that the original sources are behind a "confirmation-wall", which I until now haven't managed to circumvent. Check the segger-download some lines above, this was already a lucky draw that somebody else has already done that.

The link came btw. from [Arch's AUR](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=nrf5-sdk). But does anyone maybe like fighting the confirmation wall and help me out? In the meantime I'll add `sha256sums` for all of the downloads...

robin-nitrokey commented:
I’m a bit confused by the build profiles.  My understanding so far is:
- `nrf52` overwrites the bootloader if present and is intended for development builds without the bootloader – not needed most of the time.
- `nrf52-bootloader` is written after the bootloader, requires the bootloader and is intended for production and most development builds.

Is that correct?  Can we make this more clear – maybe `-no-bootloader` vs. `-with-bootloader` or `-testing`?

daringer commented on components/provisioner-app/Cargo.toml:
```diff
@@ -11,7 +11,8 @@ apdu-dispatch = "0.1"
 heapless-bytes = "0.3"
-lpc55-hal = { version = "0.3", features = ["littlefs", "rtic-peripherals"] }
+lpc55-hal = { version = "0.3", features = ["littlefs", "rtic-peripherals"], optional = true }
+nrf52840-pac = { version = "0.11", optional = true }
```
done

sosthene-nitrokey commented:
`make reset-nk3am.bl` gives the following error:

```
16:06:~/c/n/r/embedded/ embedded-pr>make reset
make -C ../../gen-fido-keys clean
******************************************************************************************
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[1]: *** ../../gen-fido-keys: No such file or directory.  Stop.
FAIL: var: BOARD required!!!
FAIL: var: BUILD_PROFILE required!!!
make: *** [Makefile:128: fidokeys_lfs_d8000+14000.ihex] Error 2
make: *** Waiting for unfinished jobs....
make: *** [Makefile:90: check-var-BOARD] Error 1
make: *** [Makefile:90: check-var-BUILD_PROFILE] Error 1
**** TARGET:                reset | BINARY:   artifacts/runner--.bin(.ihex)
**** BOARD:                       | SOC:      lpc55
**** PROFILE:                     | BUILD_ID: -
**** FEATURES: board- develop
******************************************************************************************
```

The correct path seems to be `../../utils/gen-fido-keys`. When fixing the path in the Makefile I still have an error:

```
rm -f lfs/lfs.a
ar qcs lfs/lfs.a lfs/lfs.o lfs/lfs_util.o
gcc -Wall -Wextra -g2 -O0 -I./lfs -DFILESYSTEM_BASE=0xd8000 -DFILESYSTEM_SIZE=0x14000 -DFILESYSTEM_BLK_SHIFT=8 gen_init_fs.c lfs/lfs.a lfs/lfs.a -o gen_init_fs
gen_init_fs.c:16:10: fatal error: my_lfs_common.h: No such file or directory
   16 | #include "my_lfs_common.h"
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [Makefile:37: gen_init_fs] Error 1
```

robin-nitrokey commented:
The `nrf-bootloader` Makefile is not sh-compatible, `[[ "4" == "4" ]]` should be `[ "4" = "4" ]`.

jnordholz commented:
@daringer This is both fixed on my branch, see 6ea0292 and 28ea851 - I guess we could just pull these to the front (i.e. this PR) and out of my on-top branch.

daringer commented:
* fixed ci builds for lpc55 runner (adapted to provisioner-app)
* removed anything that is not only building the firmware from `runners/embedded` (find convenience functions in `utils/nrf-builder`)
* rebase on top of `main´
* fix `utils/fido-gen-keys` using https://github.com/Nitrokey/nitrokey-3-firmware/commit/6ea02929390ba082224ce5183c2dee8be152d4b7 , currently there is no access point to this
* set version to 1.1.0 in `runners/embedded/Cargo.toml`
* pull in #60 into `runners/embedded/`
* removed some bashims
* tested the hell out of it
..... merge incoming
